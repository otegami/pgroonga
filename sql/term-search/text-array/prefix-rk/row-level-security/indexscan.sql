CREATE TABLE readings (
  id integer,
  user_name text,
  katakanas text[]
);

CREATE USER alice NOLOGIN;
GRANT ALL ON TABLE readings TO alice;

INSERT INTO readings VALUES
  (1, 'nonexistent', ARRAY['ポストグレスキューエル', 'ポスグレ']);
INSERT INTO readings VALUES (2, 'alice', ARRAY['グルンガ', 'グルン']);
INSERT INTO readings VALUES (3, 'alice', ARRAY['ピージールンガ', 'ピーグルン']);

ALTER TABLE readings ENABLE ROW LEVEL SECURITY;
CREATE POLICY readings_myself ON readings USING (user_name = current_user);

CREATE INDEX pgrn_index ON readings
 USING pgroonga (katakanas pgroonga_text_array_term_search_ops_v2);

SET enable_seqscan = off;
SET enable_indexscan = on;
SET enable_bitmapscan = off;

SET SESSION AUTHORIZATION alice;
\pset format unaligned
EXPLAIN (COSTS OFF)
SELECT katakanas
  FROM readings
 WHERE katakanas &^~ 'p'
 ORDER BY id
\g |sed -r -e "s/\(CURRENT_USER\)::text/CURRENT_USER/g"
\pset format aligned

SELECT katakanas
  FROM readings
 WHERE katakanas &^~ 'p'
 ORDER BY id;
RESET SESSION AUTHORIZATION;

DROP TABLE readings;

DROP USER alice;
