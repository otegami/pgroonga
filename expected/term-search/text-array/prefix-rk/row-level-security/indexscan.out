CREATE TABLE readings (
  id integer,
  user_name text,
  katakanas text[]
);
CREATE USER alice NOLOGIN;
GRANT ALL ON TABLE readings TO alice;
INSERT INTO readings VALUES
  (1, 'nonexistent', ARRAY['ポストグレスキューエル', 'ポスグレ']);
INSERT INTO readings VALUES (2, 'alice', ARRAY['グルンガ', 'グルン']);
INSERT INTO readings VALUES (3, 'alice', ARRAY['ピージールンガ', 'ピーグルン']);
ALTER TABLE readings ENABLE ROW LEVEL SECURITY;
CREATE POLICY readings_myself ON readings USING (user_name = current_user);
CREATE INDEX pgrn_index ON readings
 USING pgroonga (katakanas pgroonga_text_array_term_search_ops_v2);
SET enable_seqscan = off;
SET enable_indexscan = on;
SET enable_bitmapscan = off;
SET SESSION AUTHORIZATION alice;
\pset format unaligned
EXPLAIN (COSTS OFF)
SELECT katakanas
  FROM readings
 WHERE katakanas &^~ 'p'
 ORDER BY id
\g |sed -r -e "s/\(CURRENT_USER\)::text/CURRENT_USER/g"
QUERY PLAN
Sort
  Sort Key: id
  ->  Index Scan using pgrn_index on readings
        Index Cond: (katakanas &^~ 'p'::text)
        Filter: (user_name = CURRENT_USER)
(5 rows)
\pset format aligned
SELECT katakanas
  FROM readings
 WHERE katakanas &^~ 'p'
 ORDER BY id;
          katakanas          
-----------------------------
 {ピージールンガ,ピーグルン}
(1 row)

RESET SESSION AUTHORIZATION;
DROP TABLE readings;
DROP USER alice;
